### ISSUE TITLE
How to tell if a relationship has been loaded

### ISSUE THREAD

**Issue by ryanto** [emberjs/data#5629]
I'm wondering if Ember Data has a way to know if a relationship has already been loaded.

I'm running into a situation where reference's `load` API changes based on wether or not the relationship has been loaded.

For example...

```js
let reference = post.hasMany('comments');
reference.load() // no-ops if the comments have already been loaded
```

I'd love to have a method to check if the relationship has been loaded in Ember Data's eyes. This would allow me to do something like...

```js
let reference = post.hasMany('comments');

if (reference.hasLoaded()) {
  reference.reload();
} else {
  reference.load();
}
```

**Comment by runspired** [comment]
@ryanto we have a great deal of information about the relationship state and whether we need to fire a request encapsulated in the `relationship` state class. I've been thinking about the possibility of making this public API, because I *want* folks to write better relationship computeds (it is very doable if the state class is public).

From a private API perspective, you'll find that the relationship state class is attached to the reference. That should be enough to get you started and to figure out if there are some useful things to RFC. I suspect that proxying the method `shouldMakeRequest` and some of the other flags would make sense.

**Comment by ryanto** [comment]
Cool! I had this working with some private APIs off of the reference, but it looks like this has changed since 3.1. Previously, I was relying on `hasData` and `hasLoaded`, but these no long exist. 

That made me realize that my code is pretty brittle and I need to do this in a blessed public API way.

Given that the relationship state class isn't public, is there a way to figure this out with public APIs? Some decision tree based on reference values or ids? I've been playing around with a few things, but can't get anything to stick.

If there's no public API for this sort of thing, which it sounds like that's the case, what's the best way for me to help push one through? Addon code that leads to an RFC?

**Comment by sly7-7** [comment]
@ryanto I think you can test `reference.value()` for this ?
Maybe I don't understand what you want, but it seems to me that something like 
```
let reference = post.hasMany('comments');

if (reference.value() === null) {
  reference.load();
} else {
  reference.reload();
}
```
should work

**Comment by YoranBrondsema** [comment]
I confirm, it seems that `value() === null` for unloaded `hasMany` relationships in both 3.1 and 3.2 (I haven't tried on 3.3 nor 3.4).

That aside, I agree that it would be very helpful to have a public `hasLoaded` function.

**Comment by abdullah2891** [comment]
can we have a listener which store operation is being used on the model?
for example this.get('store').queryrecord('model'... and we can have a property which tracks it

**Comment by Herriau** [comment]
> I am surprised it works for `hasMany` relationships.

It works because `null` isn't ambiguous for `hasMany` relationships:
- `ref.value() === null` if and only if relationship is not loaded
- `ref.value() instanceof DS.ManyArray` if and only if relationship is loaded

It's ambiguous for `belongsTo` relationships because:
- `ref.value() === null`: if relationship is not loaded
- `ref.value() === null`: if relationship has loaded but its value is `null` ("optional `belongsTo`")

So we need a way to disambiguate between this two cases. The introduction of a `hasLoaded()` method on `BelongsToReference` / `HasManyReference` as suggested by @ryanto would certainly help with that.

**Comment by runspired** [comment]
@Herriau reach out on discord I'd love to pair for a little to poke at this

**Comment by joshkg** [comment]
Yearly check-in -- any progress on this issue? 🤔  Would also love to be able to check if my async relationship has been fetched or not 🙏

**Comment by runspired** [comment]
@joshkg yes, several of the linked issues have likely been resolved but need test PRs to confirm. They can be fixed fairly easily at this point if they aren't fixed already.

**Comment by runspired** [comment]
@joshkg @Herriau fixes are on master and will be in 3.28 beta 2 for meta / links / null vs undefined on belongsTo.

Closing this as this seems sufficient at this time to determine if a load has occurred.
